### PR TITLE
[Elastic Agent] Support an application spec to be ran through service manager (eg. Endpoint)

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

This splits the `pkg/core/plugin/app` into 2 distinct types of application types `process` (which runs the application as a subprocess and passes the connection credentials through stdin) and `service` (which runs the application as it is started by the service manager on the system and passes the connection credentials over a TCP socket defined in the spec file with `service: {port_number}`.

This updates the `endpoint.spec` file to set `service: 6788` so that the `pkg/core/plugin/service.Application` will start the TCP connections credentials socket on 6788.

This also moves some modules around outside of `pkg/core/plugin` to just `pkg/core` keeping `pkg/core/plugin/*` for the different application types. This also allows for better code re-use in some areas, as the 2 types have lots in common.

This currently has a pre-req on #19171, so it shows in this diff.

## Why is it important?

To enable support for Elastic Agent to control Endpoint as a service that is installed on the same machine.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- ~~[ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~~

### CODE REVIEW

File: x-pack/elastic-agent/pkg/core/plugin/process/status.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/server"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/state"
+)
+
+// OnStatusChange is the handler called by the GRPC server code.
+//
+// It updates the status of the application and handles restarting the application is needed.
+func (a *Application) OnStatusChange(s *server.ApplicationState, status proto.StateObserved_Status, msg string) {
+	a.appLock.Lock()
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/core/plugin/process/status.go:22]: we always return right before exiting, any reason why defer unlock is not preferred in this case?

File: x-pack/elastic-agent/pkg/core/plugin/service/app.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+	a.cleanUp()
+	a.stopCredsListener()
+}
+
+// OnStatusChange is the handler called by the GRPC server code.
+//
+// It updates the status of the application and handles restarting the application is needed.
+func (a *Application) OnStatusChange(s *server.ApplicationState, status proto.StateObserved_Status, msg string) {
+	a.appLock.Lock()
```
Comment (michalpristas) [inline x-pack/elastic-agent/pkg/core/plugin/service/app.go:244]: same thing with unlock

File: x-pack/elastic-agent/pkg/agent/application/managed_mode.go [hunk 3]
```diff
@@ -147,7 +146,7 @@ func newManaged(
 		router,
 		&configModifiers{
 			Decorators: []decoratorFunc{injectMonitoring},
-			Filters:    []filterFunc{filters.ConstraintFilter},
+			Filters:    []filterFunc{injectFleet(config), filters.ConstraintFilter},
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/agent/application/managed_mode.go:149]: It really is a modifier, but decorators are use after parsing the config into the spec files.

I could add a new `Modifiers` with `[]modifierFunc` if you prefer?

File: x-pack/elastic-agent/pkg/core/plugin/process/status.go (hunk 1 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/core/plugin/process/status.go:22]: updated to use `defer a.appLock.Unlock()`

File: x-pack/elastic-agent/pkg/core/plugin/service/app.go (hunk 2 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/core/plugin/service/app.go:244]: updated to use `defer a.appLock.Unlock()`

**Pull request merged**